Added support for binary_to_term(Bin, [used]).

fadushin opened the pull request:
These changes are made under the terms of the LGPL v2.1 (or any later version)
and Apache 2.0 licenses.

bettio commented on src/libAtomVM/term.h:
```diff
@@ -1121,4 +1121,26 @@ static inline int term_is_string(term t)
+static inline int term_list_member(term list, term e, Context *ctx)
+{
+    term t = list;
+    while (!term_is_nil(t)) {
```
while (term_is_nonemtpy_list()), otherwise it crashes with an improper list.

fadushin replied:
Fixed -- thank you.

bettio commented on src/libAtomVM/term.h:
```diff
@@ -1121,4 +1121,26 @@ static inline int term_is_string(term t)
+    term t = list;
+    while (!term_is_nil(t)) {
+        term *t_ptr = term_get_list_ptr(t);
+        term head = t_ptr[1];
```
Use term_get_list_head instead.

fadushin replied:
Fixed -- thank you.

bettio commented on src/libAtomVM/term.h:
```diff
@@ -1121,4 +1121,26 @@ static inline int term_is_string(term t)
+        if (term_equals(head, e, ctx)) {
+            return 1;
+        }
+        t = *t_ptr;
```
Same here, term_get_list_tail is more clear.

fadushin replied:
Fixed -- thank you.

bettio commented on src/libAtomVM/externalterm.c:
```diff
@@ -105,15 +113,15 @@ enum ExternalTermResult externalterm_from_binary(Context *ctx, term *dst, term b
     // convert
     //
-    *dst = externalterm_to_term(buf, ctx, 0);
+    *dst = externalterm_to_term_internal(buf, ctx, 0, bytes_read);
```
Shouldn't be &bytes_read?

bettio commented on src/libAtomVM/externalterm.h:
```diff
@@ -64,7 +64,7 @@ term externalterm_to_term(const void *external_term, Context *ctx, int use_heap_
  * deserialization fails.
  */
-enum ExternalTermResult externalterm_from_binary(Context *ctx, term *dst, term binary);
+enum ExternalTermResult externalterm_from_binary(Context *ctx, term *dst, term binary, size_t *bytes_read, size_t num_extra_terms);
```
I think we should document the argument that we added here, it took me a while to understand the purpose of num_extra_terms.

bettio replied:
I think we should document the arguments that we added here (`bytes_read` and `num_extra_terms`), it took me a while to understand the purpose of num_extra_terms.

fadushin commented:
Added doc @param comments in headers, to explain new parameters.  I also added a new flag when inserting an atom into the global atom table, to flag whether a copy of the input AtomString should be made.  This needs to be done in cases when atoms are instantiated from binaries in memory (as opposed to being read off mmap'd modules)

fadushin commented:
Any additional comments on this PR?  It is rebased over master and should be ready to merge.
